framework/py/flwr/serverapp/strategy/dp_adaptive_clipping.py

```diff
@@ -0,0 +1,343 @@
+# limitations under the License.
+# ==============================================================================
+"""Message-based Central differential privacy with adaptive clipping.
+
+Paper (Andrew et al.): https://arxiv.org/abs/1905.03871
+"""
+
+import math
+from abc import ABC
+from collections import OrderedDict
+from collections.abc import Iterable
+from logging import INFO, WARNING
+from typing import Optional
+
+import numpy as np
+
+from flwr.common import Array, ArrayRecord, ConfigRecord, Message, MetricRecord, log
+from flwr.common.differential_privacy import (
+    adaptive_clip_inputs_inplace,
+    add_gaussian_noise_inplace,
+    compute_adaptive_noise_params,
+    compute_stdv,
+)
+from flwr.common.differential_privacy_constants import (
+    CLIENTS_DISCREPANCY_WARNING,
+    KEY_CLIPPING_NORM,
+    KEY_NORM_BIT,
+)
+from flwr.server import Grid
+
+from .strategy import Strategy
+
+
+class DifferentialPrivacyAdaptiveBase(Strategy, ABC):
+    """Base class for DP strategies with adaptive clipping."""
+
+    # pylint: disable=too-many-arguments,too-many-instance-attributes,too-many-positional-arguments
+    def __init__(
+        self,
+        strategy: Strategy,
+        noise_multiplier: float,
+        num_sampled_clients: int,
+        initial_clipping_norm: float = 0.1,
+        target_clipped_quantile: float = 0.5,
+        clip_norm_lr: float = 0.2,
+        clipped_count_stddev: Optional[float] = None,
+    ) -> None:
+        super().__init__()
+
+        if strategy is None:
+            raise ValueError("The passed strategy is None.")
+        if noise_multiplier < 0:
+            raise ValueError("The noise multiplier should be a non-negative value.")
+        if num_sampled_clients <= 0:
+            raise ValueError(
+                "The number of sampled clients should be a positive value."
+            )
+        if initial_clipping_norm <= 0:
+            raise ValueError("The initial clipping norm should be a positive value.")
+        if not 0 <= target_clipped_quantile <= 1:
+            raise ValueError("The target clipped quantile must be in [0, 1].")
+        if clip_norm_lr <= 0:
+            raise ValueError("The learning rate must be positive.")
+        if clipped_count_stddev is not None and clipped_count_stddev < 0:
+            raise ValueError("The `clipped_count_stddev` must be non-negative.")
+
+        self.strategy = strategy
+        self.num_sampled_clients = num_sampled_clients
+        self.clipping_norm = initial_clipping_norm
+        self.target_clipped_quantile = target_clipped_quantile
+        self.clip_norm_lr = clip_norm_lr
+        (
+            self.clipped_count_stddev,
+            self.noise_multiplier,
+        ) = compute_adaptive_noise_params(
+            noise_multiplier,
+            num_sampled_clients,
+            clipped_count_stddev,
+        )
+
+    def _validate_replies(self, replies: Iterable[Message]) -> bool:
```

Comment:
i refactored the fixed-clipping strategy slightly so we can use the `validate_replies` function (no longer a method) from that module and avoid rewriting it here. Could you remove this method and use the function in the `aggregate_train` methods below?

Reply:
But is it a good practice to make this strategy dependent on the other one?

Reply:
this is more a like a utils function imo. I think it's better to avoid duplicate code if possible (easier to maintain)